### ISSUE TITLE
Autopickup items from containers

### ISSUE THREAD

**Issue by matshou** [CleverRaven/Cataclysm-DDA#54406]
### Is your feature request related to a problem? Please describe.

Autopickup is a useful (although limited) feature that allows players to auto-pickup items on tiles the player is standing on. 
Unfortunately this feature only works for items that are not in containers which makes the feature very limited.

For example if the autopickup manager had rule with value *"cigarette"* and the player walked over a zombie corpse that had a pack of cigarettes inside the pants worn by the corpse neither cigarettes or the pack of cigarettes would be picked up.

### Solution you would like.

Autopickup rules should apply to items inside containers regardless of nested depth. This would allow for much less interaction when it comes to looting corpses and other containers on the ground as items listed in the autopickup rules would be looted regardless of whether they are in a single container, a container within any other number of containers, or on the ground.

In addition to this I would like to propose the following autopickup rules:

- Pickup containers **only** under the following circumstances:
    - All items contained pass the autopickup rules.
    - The container is empty.
- Pickup items from containers **only** under the following circumstances:
    - The container contains items that **don't** match the autopickup rules.
    - The container is an electronic device and the item is a battery.
    - The item is **not** a liquid.

The rules listed above are just a starting point and should be expanded with further discussion.

### Example

Autopickup rules:

- aspirin
- light battery
- chocolate candy*
- canned tuna fish*
- codeine

Items on the ground:

- plastic prescription bottle > aspirins (10)
- pulped corpse of zombie > mp3 player (off) > light battery (5/100 battery)
- plastic bag > chocolate candy (fresh)
- small tin can > canned tuna fish (fresh)
- backpack > plastic bag > butchery refuse (2), plastic prescription bottle > codeine (10)

Expected items to be picked up:

- plastic prescription bottle > aspirins (10)
- light battery (5/100 battery)
- plastic bag > chocolate candy (fresh)
- small tin can > canned tuna fish (fresh)
- plastic prescription bottle > codeine (10)

### Describe alternatives you have considered.

No alternatives considered.

### Additional context

No additional context available.

**Comment by VoleurVolant** [comment]
My ideal conditions here are:

When a *container* is specified: 
    It should be picked up *along with* its contents by default, regardless of the weight of the contents
        If any of those contents are explicitly excluded from pickup they should be dropped
    In a perfect world, there would be a prefix within the autopickup manager like `e:gallon jug` that would allow for users to declare that they want the container to *be* emptied when it is picked up, but that may be out of scope. 

When an *item* is specified, its *container* should be picked up iff one of the following is true: 
    The item is a liquid, or (later, once powders cannot be contained loose in mesh bags) if the item is a powder
    That container is providing an extension to or stoppage of the item's spoilage time (sealed, generally) 
    The item *can* be damaged by water, and the container it is in will prevent water damage 
    The container *is not* a corpse of any kind
If the target item's container is selected for pickup and also holds other items: 
    If the container is rigid, then if any of those items are explicitly excluded from pickup, they should be dropped, and they should otherwise be kept
    If the container is non-rigid, then items that are not explicitly included for pickup should be dropped 

One unfortunate factor here is that my preferences for containers selected deliberately and containers selected by virtue of having a desired item within them are not the same - I'm not sure that can be worked around, and I'd rather sacrifice "drop non-specified items from non-rigid secondarily collected containers" than sacrifice "keep contents of targeted containers". 

I'm far from the only person who uses autopickup, so I'd like to leave things pretty open for other people to weigh in - what matters most to me is keeping bottles and jars with powders, liquids, and foods in them, and leaving behind wallets, corpses, and electronics when gathering cash cards, things dropped by zombies, and batteries. If my position is unpopular, I'm happy to live with everything else being cut.

**Comment by Terrorforge** [comment]
Duplicate of #53425

But since we're discussing the finer details: I would greatly prefer if autopickup took pills out of bottles. My ideal scenario is using pocket settings to ensure all my codeine goes in the same bottle, rather than spread out in twenty different ones. I'd even prefer this behavior for liquids in some cases, e.g. consolidating antiseptic, but that seems impractical as a general case.

Also, I just realized a problem with implementing this: your backpack almost certainly contains things you have on autopickup, so if you put it down, you'll take those things out. There needs to be some way to flag a container as "don't search this".

**Comment by matshou** [comment]
@VoleurVolant Thank you for sharing your ideas. I've implemented most of them in my PR branch #54408.

> In a perfect world, there would be a prefix within the autopickup manager like `e:gallon jug` that would allow for users to declare that they want the container to be emptied when it is picked up, but that may be out of scope.

This is an interesting idea I would like to dedicated a separate PR to since the scope of this implementation would be far bigger then just checking for a single prefix. There would need to be a system that reads flags such as `e:gallon jug` from the autopickup rules and interprets them as actions.

> The item can be damaged by water, and the container it is in will prevent water damage

I am not quite sure why you want waterproof containers to be considered for pickup.

> One unfortunate factor here is that my preferences for containers selected deliberately and containers selected by virtue of having a desired item within them are not the same - I'm not sure that can be worked around, and I'd rather sacrifice "drop non-specified items from non-rigid secondarily collected containers" than sacrifice "keep contents of targeted containers".

Could you give me a direct example of what you are talking about here?

**Comment by matshou** [comment]
> But since we're discussing the finer details: I would greatly prefer if autopickup took pills out of bottles. My ideal scenario is using pocket settings to ensure all my codeine goes in the same bottle, rather than spread out in twenty different ones. I'd even prefer this behavior for liquids in some cases, e.g. consolidating antiseptic, but that seems impractical as a general case.

@Terrorforge The changes I implemented will make the autopickup system pick items from containers and containers nested within other containers. For example if you had `aspirin*` **whitelisted** in your autopickup rules then walking over a corpse that had a backpack that contained `plastic prescription bottle > aspirin (10)` item it would result in the plastic bottle with aspirin being picked up (leaving the backpack and all other items on the ground). However if you had `plastic prescription bottle*` **blacklisted** in your autopickup rules then **only** the stack of aspirin pills would be picked up (leaving the bottle behind). Here is a visualized example of what I am talking about:

On the ground:

- zombie corpse > backpack > plastic prescription bottle > aspirin (10)
- small cardboard box > aspirin (5)
- aspirin (5)

Autopickup rules

- `aspirin*` [Include]
- `plastic prescription bottle*` [Exclude]
- `small cardboard box*` [Exclude]

Items picked up:

- aspirin (20)

> Also, I just realized a problem with implementing this: your backpack almost certainly contains things you have on autopickup, so if you put it down, you'll take those things out. There needs to be some way to flag a container as "don't search this".

The most elegant way of resolving this problem I can think of is excluding *owned* items from autopickup considerations. 
Every item picked up by the player becomes owned by automatically assigning it the player's faction. Excluding these items from consideration ensures we don't accidentally pick up our own items by walking around the base. 

I've already implemented this in 8c496a5a0ec9eda88790ccfc8306f826eac3edc5 which comes with a new game option that makes this feature configurable for anyone who still wants all items to be eligible for autopickup.
